romanisim/tests/test_linear.py

```diff
@@ -0,0 +1,144 @@
+    assert linearity.dq[22, 22] == parameters.dqbits['nonlinear']
+    # All other entries should be zero
+    assert np.count_nonzero(linearity.dq) == 2
+
+    res = linearity.apply(counts)
+
+    assert res[1, 1] == counts[1, 1]
+    assert res[22, 22] == counts[22, 22]
+    # All other entries should be the same
+    assert np.sum(res != counts) == np.prod(counts.shape) - 2
+
+def test_electrons():
+    counts = np.random.poisson(100, size=(100, 100))
+    coeffs = np.asfarray([1.0, 0.7, 3.0e-6, 5.0e-12])
+    lin_coeffs = np.tile(coeffs[:, np.newaxis, np.newaxis], (1, 100, 100))
+    lin_coeffs[:, 0:50, :] *= 2.0
+    gain = 4.0 * u.electron / u.DN
+
+    linearity = nonlinearity.NL(lin_coeffs, gain)
+
+    res = linearity.apply(counts)
+
+    res_elec = linearity.apply(gain * counts, electrons=True)
+
+    assert np.all(res_elec[:] == gain * res[:])
+    assert res_elec.unit == u.electron / u.DN
+    assert not hasattr(res, "unit")
+
+def test_reverse():
+    counts = np.random.poisson(100, size=(100, 100))
+    coeffs = np.asfarray([1.0, 0.7, 3.0e-6, 5.0e-12])
+    lin_coeffs = np.tile(coeffs[:, np.newaxis, np.newaxis], (1, 100, 100))
+
+    lin_coeffs[:, 0:50, :] *= 2.0
+    rev_lin_coeffs = lin_coeffs[::-1, ...]
+    gain = 4.0 * u.electron / u.DN
+
+    linearity = nonlinearity.NL(lin_coeffs, gain)
+    rev_linearity = nonlinearity.NL(rev_lin_coeffs, gain)
+
+    res = linearity.apply(counts)
+    res_rev = rev_linearity.apply(counts, reversed=True)
+
+    assert np.all(res_rev[:] == res[:])
+
+@pytest.mark.skipif(
+    os.environ.get("CI") == "true",
+    reason=(
+        "Roman CRDS servers are not currently available outside the internal network"
+    ),
+)
+def test_inverse_then_linearity():
+    # Test that applying inverse linearity and then linearity returns the results to
+    # the original value
+
+    reffiles = crds.getreferences(
+        {'roman.meta.instrument.name': 'WFI',
+         'roman.meta.instrument.detector':'WFI01',
+         'roman.meta.exposure.start_time': '2026-01-01T00:00:00'},
+        reftypes=['inverselinearity', 'linearity'],
+        observatory='roman')
+
+    inverse_linearity_model = roman_datamodels.datamodels.InverselinearityRefModel(
+        reffiles['inverselinearity'])
+    linearity_model = roman_datamodels.datamodels.LinearityRefModel(
+        reffiles['linearity'])
+
+    inverse_linearity = nonlinearity.NL(
+        inverse_linearity_model.coeffs[:,4:-4,4:-4], gain=1.0)
+    linearity = nonlinearity.NL(
+        linearity_model.coeffs[:,4:-4,4:-4], gain=1.0)
+
+    counts = np.random.poisson(16000, size=(4088, 4088))
+    level_0_inv = inverse_linearity.apply(counts)
+    level_0_lin = linearity.apply(level_0_inv)
+
+    # Divide initial counts before and after inverse & NL application
+    div = (counts[np.isfinite(counts)] / level_0_lin[np.isfinite(level_0_lin)]).flatten()
+
+    assert np.isclose(np.average(div), 1.0, atol=0.1)
```

Comment:
It doesn't look like these were very close?  I think the inverse linearity is literally intended to be the inverse of the linearity, so I was hoping for pretty good agreement.  What atol / rtol is needed if you wanted the test to be
np.isclose(counts[m], level_0_lin[m])?

Reply:
I just pushed an update here that looks pretty good to me.  For the ~50% of pixels that aren't obviously bonkers, the corrections look pretty good, and invert at the 1e-6 level at your chosen flux level.  I've put in a fractional uncertainty rms of 1e-5.

Reply:
Okay, I have downloaded it and merged it with my local work.